src/plugins/intel_cpu/src/nodes/subgraph.cpp

```diff
@@ -43,63 +46,207 @@ namespace intel_cpu {
         return EMPTY_PORT_MASK;
     }
 
 private:
-    Snippet* m_node;
+    std::shared_ptr<ov::Model> m_body;
 };
 
 class SnippetShapeInferFactory : public ShapeInferFactory {
 public:
-    SnippetShapeInferFactory(Snippet* node) : m_node(node) {}
+    SnippetShapeInferFactory(const std::shared_ptr<ov::Node>& op) {
+        auto subgraph = ov::as_type_ptr<snippets::op::Subgraph>(op);
+        snippet_body = subgraph->body_ptr()->clone();
+    }
     ShapeInferPtr makeShapeInfer() const override {
-        return std::make_shared<SnippetShapeInfer>(m_node);
+        return std::make_shared<SnippetShapeInfer>(snippet_body);
     }
 
 private:
-    Snippet* m_node;
+    std::shared_ptr<ov::Model> snippet_body = nullptr;
 };
 } // namespace
 
 Snippet::Snippet(const std::shared_ptr<ov::Node>& op, const GraphContext::CPtr context)
-        : Node(op, context, SnippetShapeInferFactory(this)) {
+        : Node(op, context, SnippetShapeInferFactory(op)) {
     host_isa = dnnl::impl::cpu::x64::mayiuse(dnnl::impl::cpu::x64::avx512_core) ?
         dnnl::impl::cpu::x64::avx512_core : dnnl::impl::cpu::x64::avx2;
     original_snippet = ov::as_type_ptr<snippets::op::Subgraph>(op);
     if (!original_snippet) {
         IE_THROW(NotImplemented) << "Node is not an instance of snippets::op::Subgraph";
     }
+    init_body_hash();
+    is_dynamic = isDynamicNgraphNode(op);
 }
 
-void Snippet::copy_snippet() {
+void Snippet::copy_snippet() const {
     ov::OutputVector subgraph_node_inputs;
     for (const auto &input : original_snippet->input_values()) {
         auto new_input = std::make_shared<ov::opset1::Parameter>(input.get_element_type(), input.get_partial_shape());
         subgraph_node_inputs.push_back(new_input);
     }
     std::shared_ptr<ov::Model> new_body = original_snippet->body_ptr()->clone();
-    snippet = std::make_shared<snippets::op::Subgraph>(subgraph_node_inputs, new_body);
-    ov::copy_runtime_info(original_snippet, snippet);
-    snippet->set_friendly_name(original_snippet->get_friendly_name());
-#if defined(OPENVINO_ARCH_X86_64)
-    snippet->set_generator(std::make_shared<CPUGenerator>(host_isa));
-    isa_num_lanes =  snippet->get_generator()->get_target_machine()->get_lanes();
-#else
-    IE_THROW(NotImplemented) << "CPU plugin: code-generation is not supported on non-x64 platforms";
-#endif // OPENVINO_ARCH_X86_64
+    snippetAttrs.snippet = std::make_shared<snippets::op::Subgraph>(subgraph_node_inputs, new_body);
+    ov::copy_runtime_info(original_snippet, snippetAttrs.snippet);
+    snippetAttrs.snippet->set_friendly_name(original_snippet->get_friendly_name());
+    snippetAttrs.snippet->set_generator(std::make_shared<CPUGenerator>(host_isa));
+}
+
+void Snippet::init_body_hash() {
+    uint64_t seed = 0;
+    ov::pass::Manager m;
+    m.register_pass<ov::pass::Hash>(seed);
+    m.run_passes(std::const_pointer_cast<ov::Model>(original_snippet->body_ptr()));
+    snippetAttrs.bodyHash = seed;
```

Comment:
Maybe it would be better
```suggestion
    uint64_t seed = 0;
    ov::pass::Hash hash_function(seed); 
    hash_function.run_on_model(original_snippet->body_ptr());
    snippetAttrs.bodyHash = seed;
```

Moreover as I know, to use `std::const_pointer_cast` is not good practice. So let's try to avoid this using if possible 🙂

Reply:
yes, applied, thanks Alexandra!